File: 2025/go/error_handling.py
def read_and_sum_integers(file_path):
    try:
        with open(file_path, "r", encoding="utf-8") as file:
            total = 0
            for line in file:
                try:
                    num = int(line.strip())
                    total += num
                except ValueError as e:
                    raise ValueError(
                        f"Invalid content in file: '{line.strip()}'"
                    ) from e
            return total
    except FileNotFoundError as e:
        raise FileNotFoundError(f"File not found: {file_path}") from e
    except OSError as e:
        raise RuntimeError(f"An unexpected OS error occurred: {e}") from e


def main():
    file_path = "numbers.txt"

    try:
        total = read_and_sum_integers(file_path)
        print(f"Sum of integers: {total}")
    except (FileNotFoundError, ValueError, RuntimeError) as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()

File: 2025/go/error_handling.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Custom error for invalid content
var ErrInvalidContent = errors.New("file contains invalid content")

// Function to read a file and parse integers
func readAndSumIntegers(filePath string) (int, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return 0, fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	sum := 0
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		num, err := strconv.Atoi(line)
		if err != nil {
			return 0, fmt.Errorf("%w: '%s'", ErrInvalidContent, line)
		}
		sum += num
	}

	if err := scanner.Err(); err != nil {
		return 0, fmt.Errorf("error reading file: %w", err)
	}

	return sum, nil
}

func main() {
	filePath := "numbers.txt"

	// Call the function and handle errors
	sum, err := readAndSumIntegers(filePath)
	if err != nil {
		if errors.Is(err, ErrInvalidContent) {
			fmt.Printf("Content error: %v\\n", err)
		} else {
			fmt.Printf("General error: %v\\n", err)
		}
		return
	}

	fmt.Printf("Sum of integers: %d\\n", sum)
}
